File: src/carbonfactor_parser/persistence/postgresql_schema_bootstrap.py
"""Runtime-passive PostgreSQL schema bootstrap boundary contracts."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from carbonfactor_parser.persistence.postgresql_schema_catalog import (
    get_required_table_names,
)


class PostgreSQLSchemaBootstrapMode(str, Enum):
    """Caller intent for future PostgreSQL schema bootstrap handling."""

    CHECK_ONLY = "check_only"
    CREATE_MISSING = "create_missing"


class PostgreSQLSchemaBootstrapTableStatus(str, Enum):
    """Per-table status values for schema bootstrap reports."""

    REQUIRED = "required"
    PRESENT = "present"
    MISSING = "missing"
    CREATED = "created"
    SKIPPED = "skipped"


@dataclass(frozen=True)
class PostgreSQLSchemaBootstrapRequest:
    """Schema bootstrap intent metadata without runtime execution."""

    mode: PostgreSQLSchemaBootstrapMode
    required_table_names: tuple[str, ...]
    fail_on_missing: bool = True


@dataclass(frozen=True)
class PostgreSQLSchemaBootstrapTableResult:
    """Reported status for one required schema table."""

    table_name: str
    status: PostgreSQLSchemaBootstrapTableStatus
    reason: str = ""


@dataclass(frozen=True)
class PostgreSQLSchemaBootstrapReport:
    """Runtime-passive schema bootstrap request report."""

    mode: PostgreSQLSchemaBootstrapMode
    required_table_names: tuple[str, ...]
    table_results: tuple[PostgreSQLSchemaBootstrapTableResult, ...]
    fail_on_missing: bool
    no_execution: bool
    opens_connection: bool
    runs_sql: bool
    creates_tables_now: bool
    runs_migrations: bool
    reads_environment: bool
    writes_files: bool
    performs_network_calls: bool

    @property
    def missing_table_names(self) -> tuple[str, ...]:
        return tuple(
            result.table_name
            for result in self.table_results
            if result.status is PostgreSQLSchemaBootstrapTableStatus.MISSING
        )

    @property
    def created_table_names(self) -> tuple[str, ...]:
        return tuple(
            result.table_name
            for result in self.table_results
            if result.status is PostgreSQLSchemaBootstrapTableStatus.CREATED
        )

    @property
    def present_table_names(self) -> tuple[str, ...]:
        return tuple(
            result.table_name
            for result in self.table_results
            if result.status is PostgreSQLSchemaBootstrapTableStatus.PRESENT
        )

    @property
    def skipped_table_names(self) -> tuple[str, ...]:
        return tuple(
            result.table_name
            for result in self.table_results
            if result.status is PostgreSQLSchemaBootstrapTableStatus.SKIPPED
        )


def build_postgresql_phase1_schema_bootstrap_request(
    mode: PostgreSQLSchemaBootstrapMode | str = (
        PostgreSQLSchemaBootstrapMode.CHECK_ONLY
    ),
    *,
    fail_on_missing: bool = True,
) -> PostgreSQLSchemaBootstrapRequest:
    """Build a deterministic Phase 1 schema bootstrap request."""

    return PostgreSQLSchemaBootstrapRequest(
        mode=PostgreSQLSchemaBootstrapMode(mode),
        required_table_names=get_required_table_names(),
        fail_on_missing=fail_on_missing,
    )


def build_postgresql_phase1_schema_bootstrap_report(
    mode: PostgreSQLSchemaBootstrapMode | str = (
        PostgreSQLSchemaBootstrapMode.CHECK_ONLY
    ),
    *,
    present_table_names: tuple[str, ...] = (),
    created_table_names: tuple[str, ...] = (),
    skipped_table_names: tuple[str, ...] = (),
    fail_on_missing: bool = True,
) -> PostgreSQLSchemaBootstrapReport:
    """Build a passive report for Phase 1 schema bootstrap status metadata."""

    request = build_postgresql_phase1_schema_bootstrap_request(
        mode=mode,
        fail_on_missing=fail_on_missing,
    )
    present_tables = set(present_table_names)
    created_tables = set(created_table_names)
    skipped_tables = set(skipped_table_names)

    table_results = tuple(
        PostgreSQLSchemaBootstrapTableResult(
            table_name=table_name,
            status=_resolve_table_status(
                table_name=table_name,
                mode=request.mode,
                present_table_names=present_tables,
                created_table_names=created_tables,
                skipped_table_names=skipped_tables,
            ),
            reason=_resolve_table_reason(
                table_name=table_name,
                mode=request.mode,
                present_table_names=present_tables,
                created_table_names=created_tables,
                skipped_table_names=skipped_tables,
            ),
        )
        for table_name in request.required_table_names
    )

    return PostgreSQLSchemaBootstrapReport(
        mode=request.mode,
        required_table_names=request.required_table_names,
        table_results=table_results,
        fail_on_missing=request.fail_on_missing,
        no_execution=True,
        opens_connection=False,
        runs_sql=False,
        creates_tables_now=False,
        runs_migrations=False,
        reads_environment=False,
        writes_files=False,
        performs_network_calls=False,
    )


def _resolve_table_status(
    *,
    table_name: str,
    mode: PostgreSQLSchemaBootstrapMode,
    present_table_names: set[str],
    created_table_names: set[str],
    skipped_table_names: set[str],
) -> PostgreSQLSchemaBootstrapTableStatus:
    if table_name in present_table_names:
        return PostgreSQLSchemaBootstrapTableStatus.PRESENT
    if (
        mode is PostgreSQLSchemaBootstrapMode.CREATE_MISSING
        and table_name in created_table_names
    ):
        return PostgreSQLSchemaBootstrapTableStatus.CREATED
    if table_name in skipped_table_names:
        return PostgreSQLSchemaBootstrapTableStatus.SKIPPED
    return PostgreSQLSchemaBootstrapTableStatus.MISSING


def _resolve_table_reason(
    *,
    table_name: str,
    mode: PostgreSQLSchemaBootstrapMode,
    present_table_names: set[str],
    created_table_names: set[str],
    skipped_table_names: set[str],
) -> str:
    if table_name in present_table_names:
        return "Required table was reported present by caller metadata."
    if (
        mode is PostgreSQLSchemaBootstrapMode.CREATE_MISSING
        and table_name in created_table_names
    ):
        return "Required table was reported created by caller metadata."
    if table_name in skipped_table_names:
        return "Required table was reported skipped by caller metadata."
    return "Required table was not reported present or created."

File: src/dotnet/CarbonOps.Parser.Contracts/PostgreSQLSchemaBootstrap.cs
namespace CarbonOps.Parser.Contracts;

public enum PostgreSQLSchemaBootstrapMode
{
    CheckOnly = 0,
    CreateMissing = 1,
}

public enum PostgreSQLSchemaBootstrapTableStatus
{
    Required = 0,
    Present = 1,
    Missing = 2,
    Created = 3,
    Skipped = 4,
}

public sealed record PostgreSQLSchemaBootstrapRequest
{
    public PostgreSQLSchemaBootstrapMode Mode { get; }

    public IReadOnlyList<string> RequiredTableNames { get; }

    public bool FailOnMissing { get; }

    public PostgreSQLSchemaBootstrapRequest(
        PostgreSQLSchemaBootstrapMode mode,
        IEnumerable<string> requiredTableNames,
        bool failOnMissing = true)
    {
        Mode = mode;
        RequiredTableNames = Array.AsReadOnly(requiredTableNames.ToArray());
        FailOnMissing = failOnMissing;
    }
}

public sealed record PostgreSQLSchemaBootstrapTableResult(
    string TableName,
    PostgreSQLSchemaBootstrapTableStatus Status,
    string Reason = "");

public sealed record PostgreSQLSchemaBootstrapReport
{
    public PostgreSQLSchemaBootstrapMode Mode { get; }

    public IReadOnlyList<string> RequiredTableNames { get; }

    public IReadOnlyList<PostgreSQLSchemaBootstrapTableResult> TableResults { get; }

    public bool FailOnMissing { get; }

    public bool NoExecution { get; }

    public bool OpensConnection { get; }

    public bool RunsSql { get; }

    public bool CreatesTablesNow { get; }

    public bool RunsMigrations { get; }

    public bool ReadsEnvironment { get; }

    public bool WritesFiles { get; }

    public bool PerformsNetworkCalls { get; }

    public IReadOnlyList<string> MissingTableNames => Array.AsReadOnly(
        TableResults
            .Where(result => result.Status == PostgreSQLSchemaBootstrapTableStatus.Missing)
            .Select(result => result.TableName)
            .ToArray());

    public IReadOnlyList<string> PresentTableNames => Array.AsReadOnly(
        TableResults
            .Where(result => result.Status == PostgreSQLSchemaBootstrapTableStatus.Present)
            .Select(result => result.TableName)
            .ToArray());

    public IReadOnlyList<string> CreatedTableNames => Array.AsReadOnly(
        TableResults
            .Where(result => result.Status == PostgreSQLSchemaBootstrapTableStatus.Created)
            .Select(result => result.TableName)
            .ToArray());

    public IReadOnlyList<string> SkippedTableNames => Array.AsReadOnly(
        TableResults
            .Where(result => result.Status == PostgreSQLSchemaBootstrapTableStatus.Skipped)
            .Select(result => result.TableName)
            .ToArray());

    public PostgreSQLSchemaBootstrapReport(
        PostgreSQLSchemaBootstrapMode mode,
        IEnumerable<string> requiredTableNames,
        IEnumerable<PostgreSQLSchemaBootstrapTableResult> tableResults,
        bool failOnMissing,
        bool noExecution,
        bool opensConnection,
        bool runsSql,
        bool createsTablesNow,
        bool runsMigrations,
        bool readsEnvironment,
        bool writesFiles,
        bool performsNetworkCalls)
    {
        Mode = mode;
        RequiredTableNames = Array.AsReadOnly(requiredTableNames.ToArray());
        TableResults = Array.AsReadOnly(tableResults.ToArray());
        FailOnMissing = failOnMissing;
        NoExecution = noExecution;
        OpensConnection = opensConnection;
        RunsSql = runsSql;
        CreatesTablesNow = createsTablesNow;
        RunsMigrations = runsMigrations;
        ReadsEnvironment = readsEnvironment;
        WritesFiles = writesFiles;
        PerformsNetworkCalls = performsNetworkCalls;
    }
}

public static class PostgreSQLSchemaBootstrapBoundary
{
    public static IReadOnlyList<string> RequiredPhase1TableNames { get; } = Array.AsReadOnly(
        PostgreSQLRuntimeSchemaCatalog.RequiredTableNames.ToArray());

    public static PostgreSQLSchemaBootstrapRequest CreateRequest(
        PostgreSQLSchemaBootstrapMode mode = PostgreSQLSchemaBootstrapMode.CheckOnly,
        bool failOnMissing = true) =>
        new(mode, RequiredPhase1TableNames, failOnMissing);

    public static PostgreSQLSchemaBootstrapReport BuildReport(
        PostgreSQLSchemaBootstrapMode mode = PostgreSQLSchemaBootstrapMode.CheckOnly,
        IEnumerable<string>? presentTableNames = null,
        IEnumerable<string>? createdTableNames = null,
        IEnumerable<string>? skippedTableNames = null,
        bool failOnMissing = true)
    {
        var request = CreateRequest(mode, failOnMissing);
        var presentTables = (presentTableNames ?? []).ToHashSet(StringComparer.Ordinal);
        var createdTables = (createdTableNames ?? []).ToHashSet(StringComparer.Ordinal);
        var skippedTables = (skippedTableNames ?? []).ToHashSet(StringComparer.Ordinal);

        var tableResults = request.RequiredTableNames.Select(tableName => new PostgreSQLSchemaBootstrapTableResult(
            tableName,
            ResolveTableStatus(tableName, request.Mode, presentTables, createdTables, skippedTables),
            ResolveTableReason(tableName, request.Mode, presentTables, createdTables, skippedTables)));

        return new PostgreSQLSchemaBootstrapReport(
            request.Mode,
            request.RequiredTableNames,
            tableResults,
            request.FailOnMissing,
            noExecution: true,
            opensConnection: false,
            runsSql: false,
            createsTablesNow: false,
            runsMigrations: false,
            readsEnvironment: false,
            writesFiles: false,
            performsNetworkCalls: false);
    }

    private static PostgreSQLSchemaBootstrapTableStatus ResolveTableStatus(
        string tableName,
        PostgreSQLSchemaBootstrapMode mode,
        ISet<string> presentTableNames,
        ISet<string> createdTableNames,
        ISet<string> skippedTableNames)
    {
        if (presentTableNames.Contains(tableName))
        {
            return PostgreSQLSchemaBootstrapTableStatus.Present;
        }

        if (mode == PostgreSQLSchemaBootstrapMode.CreateMissing &&
            createdTableNames.Contains(tableName))
        {
            return PostgreSQLSchemaBootstrapTableStatus.Created;
        }

        if (skippedTableNames.Contains(tableName))
        {
            return PostgreSQLSchemaBootstrapTableStatus.Skipped;
        }

        return PostgreSQLSchemaBootstrapTableStatus.Missing;
    }

    private static string ResolveTableReason(
        string tableName,
        PostgreSQLSchemaBootstrapMode mode,
        ISet<string> presentTableNames,
        ISet<string> createdTableNames,
        ISet<string> skippedTableNames)
    {
        if (presentTableNames.Contains(tableName))
        {
            return "Required table was reported present by caller metadata.";
        }

        if (mode == PostgreSQLSchemaBootstrapMode.CreateMissing &&
            createdTableNames.Contains(tableName))
        {
            return "Required table was reported created by caller metadata.";
        }

        if (skippedTableNames.Contains(tableName))
        {
            return "Required table was reported skipped by caller metadata.";
        }

        return "Required table was not reported present or created.";
    }
}
